### ISSUE TITLE
gce module ignores specified service account, always uses project default

### ISSUE THREAD

**Issue by ghost** [ansible/ansible#21122]
(resubmit of https://github.com/ansible/ansible-modules-core/issues/5376 after verifying the issue still exists in 2.2.1.0)

##### ISSUE TYPE

- Bug Report
##### COMPONENT NAME

gce
##### ANSIBLE VERSION

```

ansible 2.2.1.0
  config file = /home/gcetest/.ansible.cfg
  configured module search path = Default w/o overrides

```
##### CONFIGURATION

~/.ansible.cfg:

```
[defaults]
inventory = $HOME/.ansible.hosts
host_key_checking = False
display_skipped_hosts = False
display_args_to_stdout = False
nocows = 1
retry_files_enabled = False
module_compression = 'ZIP_STORED'

[ssh_connection]
pipelining = True
scp_if_ssh = True
```

~/.ansible.hosts:

```
localhost
```
##### OS / ENVIRONMENT

`ubuntu-14.04` image running on existing GCE node, with `libcloud-1.4.0` installed via pip
(note: this also requires `backports.ssl-match-hostname` to function, an undeclared dependency)
##### SUMMARY

Ansible's "gce" module _always_ creates the spawned machine with the default service account for our project.  The specified service account in `service_account_email` is NOT assigned to the new node.  

Note that `service_account_permissions` _does_ seem to be used because originally we tried to run an app on a created node and it didn't have the needed permissions; changing the `service_account_permissions` to include the ones required (in our case bigquery) did work.
##### STEPS TO REPRODUCE

1. create a service account that has all possible assigned roles (so permissions aren't an issue)
2. manually create a GCE instance to act as Ansible controller using above service account
3. verify that the instance has the correct service account in web UI
4. install `ansible` on the instance via apt repository (see version above)
5. install `libcloud` via `pip` (see version above)
6. run ansible gce module play (see below)

here's the task running on the GCE-based Ansible controller that's making the instance:

```
    - name: create_gce_instance                                                                                                       
      gce:                                                                                                                            

        name: '{{instance_name}}'                                                                                                     
        zone: '{{gce_compute_zone}}'                                                                                                  
        image: '{{worker_machine_image}}'                                                                                             
        machine_type: '{{worker_machine_type}}'                                                                                       

        project_id: '{{gce_project}}'                                                                                                 
        credentials_file: '{{ctlr_service_creds}}'                                                                                    
        service_account_email: '{{ctlr_service_acct}}'                                                                                
        service_account_permissions:                                                                                                  
          - bigquery                                                                                                                  
          - cloud-platform                                                                                                            
          - compute-ro                                                                                                                
          - useraccounts-ro                                                                                                           
          - logging-write                                                                                                             
          - storage-rw                                                                                                                

        # inject metadata from extra args                                                                                             
        metadata:                                                                                                                     
          source_path: '{{source_path}}'                                                                                              
          pubsub_id: '{{pubsub_id}}'                                                                                                  

      register: etlnode
```

This is running with `hosts: localhost`, `connection: local`, `gather_facts: no`, and `become: False`

Note that the `credentials_file` contains a JSON format keyfile downloaded in the web UI for the same service account specified in `service_account_email`.  I assume that the `credentials_file` gives the credentials used to create the machine, while the `service_account_email` contains the desired service account the created machine should get (since `service_account_permissions` does that for access scopes, i.e. specifies what you want the new instance to have).  We're using the same one in both just for testing because it has every permission, but eventually of course want to use a reduced permission set service account for the spawned worker. 

##### EXPECTED RESULTS

created instance will have the service account specified in `service_account_email`
##### ACTUAL RESULTS

created node _always_ has our project's default service account, despite `service_account_email` being used in the play

**Comment by ansibot** [comment]
cc @supertom
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ghost** [comment]
I tried this just now to verify the issue still exists with current stable branch: same thing. My libcloud is 1.4.0.

We're iterating on our GCE stuff now, so I'm revisiting this issue and looking to break out permissions, so far every node will have to be created with the default service account, and we have to use these "scopes" instead (i.e. `service_account_permissions`) which are (1) limited; (2) course-grained; (3) described as deprecated in Google documentation (it says to use service accounts, which we cannot because of this issue).

Based on the discussion in the original bug, it seems that we want to add another parameter instead of using `service_account_email` so as not to break existing users (even though this field essentially does nothing if present, right?), but I looked through the `devel` trunk, and I don't see any such new parameter.

I'll also note that https://github.com/ansible/ansible-modules-core/pull/5395 appears to have been one approach to solve this problem.  There was discussion, but I don't see a resolution.

I looked through both opened and closed tickets in this repository and couldn't find any.  Hence my resubmission.  Thanks.

**Comment by ghost** [comment]
updated version numbers in the bug text (ansible-2.2.1.0 and apache-libcloud-1.4.0)

**Comment by abadger** [comment]
Thanks @scottingly!  Will wait on @supertom, the maintainer of this module, to take a look.

**Comment by ghost** [comment]
Are there any plans to fix this? For the latest example, we need machines to provision at GCE using images we bake, so we need to assign `ImageUser` roles to nodes for access to the images from outside the project.  `ImageUser` is not in the default service account and cannot be added.  Already we have to separate out into different projects where we normally would not because we can't use roles to grant different capabilities.  `ImageUser` is not in the list of scope that can be assigned either, so we can't just use `service_account_permissions` to work around this.

**Comment by ghost** [comment]
actually it turns out, I can assign `roles.imageUser` on the account that binds to create the instance, by adding that service account (the one for which creds file given at instance create time) to the image bake project, so my latest use case doesn't apply.

I can still carry credentials in files and use different projects and network for isolation to work around not having roles, until the `service_account_email` or other newly-named parameter is implemented.

**Comment by ansibot** [comment]
cc @erjohnso
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by jayme-github** [comment]
I just tried something like this: https://github.com/traum-ferienwohnungen/ansible/commit/74722becf2fc65fa0304aadd90e0f1fab6aa3503

But that had no effect on the created instance (and I don't know how to debug that)

**Comment by erjohnso** [comment]
This will first need to get addressed in [libcloud](https://github.com/apache/libcloud). IIRC, the libcloud tried to support a list of service accounts (seems to be supported in the JSON request schema), but in practice only once service account is allowed (and likely uses the `default` as a fallback).

**Comment by nhammond** [comment]
It looks like libcloud will at least allow a single non-default service account now. Merged into trunk, expected in the next release. (Not yet released as of apache-libcloud==2.2.1) https://github.com/apache/libcloud/commit/11337e2020839a65e0f0102336b614e56d32549b

Not sure about support for multiple service accounts, but even allowing 1 service account in ansible would handle most use cases.

**Comment by ansibot** [comment]
cc @rambleraptor
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by rambleraptor** [comment]
Hello! The new `gcp_compute_instance` module is replacing the `gce` module. This module uses a specified service account (or you could use the default service account).

close_me
